### PR TITLE
CORE: Option to disable lock free MT queue

### PR DESCRIPTION
**Pull request by bureddy**

### CODE REVIEW

File: src/core/ucc_progress_queue_st.c [hunk 1]
```diff
@@ -64,6 +64,7 @@ ucc_status_t ucc_pq_st_init(ucc_progress_queue_t **pq)
     }
     ucc_list_head_init(&pq_st->list);
     pq_st->super.enqueue  = ucc_pq_st_enqueue;
+    pq_st->super.dequeue = NULL;
```
Comment (vspetrov) [inline src/core/ucc_progress_queue_st.c:67]: align

File: src/core/ucc_progress_queue_mt.c [hunk 2]
```diff
@@ -35,15 +44,32 @@ static void ucc_pq_mt_enqueue(ucc_progress_queue_t *pq, ucc_coll_task_t *task)
             return;
         }
     }
+
+    ucc_pq_mt_enqueue(pq, task);
+}
+
+inline static void ucc_pq_mt_dequeue(ucc_progress_queue_t *pq,
```
Comment (vspetrov) [inline src/core/ucc_progress_queue_mt.c:51]: is it really inline? you pass it as fn pointer on pq->dequeue, can it be inline this way?

File: src/core/ucc_progress_queue_mt.c [hunk 3]
```diff
@@ -35,15 +44,32 @@ static void ucc_pq_mt_enqueue(ucc_progress_queue_t *pq, ucc_coll_task_t *task)
...
+                                     ucc_coll_task_t     **popped_task_ptr,
+                                     int                   is_first_call)
+{
+    ucc_pq_mt_t *pq_mt      = ucc_derived_of(pq, ucc_pq_mt_t);
```
Comment (vspetrov) [inline src/core/ucc_progress_queue_mt.c:55]: align on = sign

File: src/core/ucc_progress_queue_mt.c [hunk 4]
```diff
@@ -24,6 +24,15 @@ typedef struct ucc_pq_mt {
 } ucc_pq_mt_t; // TODO the struct isn't a queue because not maintaining order, maybe another name
 
 static void ucc_pq_mt_enqueue(ucc_progress_queue_t *pq, ucc_coll_task_t *task)
+{
+    ucc_pq_mt_t *pq_mt      = ucc_derived_of(pq, ucc_pq_mt_t);
```
Comment (vspetrov) [inline src/core/ucc_progress_queue_mt.c:28]: extra space after pq_mt

File: src/core/ucc_context.c [hunk 5]
```diff
@@ -18,6 +18,10 @@ static ucc_config_field_t ucc_context_config_table[] = {
      "An optimization hint of how many endpoints will be created on this context",
      ucc_offsetof(ucc_context_config_t, estimated_num_eps), UCC_CONFIG_TYPE_UINT},
 
+    {"MT_LOCK_FREE_PROGRESS_Q", "1",
```
Comment (vspetrov) [inline src/core/ucc_context.c:21]: set it to 0 by default
we will enable it when we figure out where perf limitations come from

File: src/core/ucc_progress_queue_st.c (hunk 1 above)
Comment (manjugv) [inline src/core/ucc_progress_queue_st.c:67]: align

File: src/core/ucc_context.c (hunk 5 above)
Comment (manjugv) [inline src/core/ucc_context.c:21]: Naming: do we need MT in the name?

File: src/core/ucc_progress_queue_mt.c (hunk 2 above)
Comment (bureddy) [reply src/core/ucc_progress_queue_mt.c]: I dont know. removed inline

File: src/core/ucc_context.c (hunk 5 above)
Comment (bureddy) [reply src/core/ucc_context.c]: lock free is only in MT.

**Review by vspetrov — APPROVED** [review]

File: src/core/ucc_context.c (hunk 5 above)
Comment (manjugv) [reply src/core/ucc_context.c]: Yes, that is why it is redundant. It is implied. :)

Comment (bureddy) [reply src/core/ucc_context.c]: fixed

**Review by manjugv — APPROVED** [review]

**Pull request merged**